### ISSUE TITLE
Add detection/prevention of secret leaks

### ISSUE THREAD

**Issue by yotommy** [civiform/civiform#7099]
#### Is your feature request related to a problem? Please describe.

It is easy to inadvertently commit a secret (server secret, auth secret, etc.) that is used for developer testing. We would like to be notified if this occurs, or better, we could prevent it from happening in the first place.

#### What priority is this?

This is relatively low priority: no civic entity has requested this, and it doesn't block any other work.

#### Done when:

What conditions need to be met before this issue can be closed?

Tool(s) added to our toolchain to detect, and optionally prevent, secrets being committed to our repos.

**Comment by yotommy** [comment]
There are commercial solutions such as [GitGuardian](https://www.gitguardian.com/) and [GitHub Advanced Security](https://docs.github.com/en/get-started/learning-about-github/about-github-advanced-security).

However, there are also popular open-source solutions like [Gitleaks](https://github.com/gitleaks/gitleaks) (15.1k stars) and [TruffleHog](https://trufflesecurity.com/trufflehog) (13.7k stars) as well. Both of these support pre-commit hooks for prevention as well as integration with Github actions for detection.

**Comment by yotommy** [comment]
Both gitleaks and pre-commit have support in codespaces. See #7176 for WIP.

**Comment by yotommy** [comment]
@gwendolyngoetz discovered that [GitHub Advanced Security](https://docs.github.com/en/get-started/learning-about-github/about-github-advanced-security) supports [secret scanning](https://docs.github.com/en/code-security/secret-scanning/configuring-secret-scanning-for-your-repositories) and [push protection](https://docs.github.com/en/code-security/secret-scanning/push-protection-for-repositories-and-organizations) from both the web UI and the CLI.

If a push is blocked, the user can access a provided GitHub URL to override the protection. GitHub creates an [alert](https://docs.github.com/en/code-security/secret-scanning/push-protection-for-repositories-and-organizations#about-push-protection-for-repositories-and-organizations) in such cases.

The [documentation](https://docs.github.com/en/get-started/learning-about-github/about-github-advanced-security) states:

> Some features of GitHub Advanced Security are also available for public repositories on GitHub.com. For more information, see "[GitHub’s plans](https://docs.github.com/en/get-started/learning-about-github/githubs-plans)."

The same page shows a table that indicates that secret scanning is available for public repositories. We should test and see if push protection is included.

![image](https://github.com/civiform/civiform/assets/1391583/2de22267-76b9-4229-aa75-6be0840d9b7d)

**Comment by yotommy** [comment]
For testing, I created the [secret-testing repo](https://github.com/civiform/secret-testing).

I tried pushing a commit containing the first entry (`adafruit_io_key`) on the list of [supported secrets](https://docs.github.com/en/enterprise-cloud@latest/code-security/secret-scanning/secret-scanning-patterns#supported-secrets). The [push](https://github.com/civiform/secret-testing/commit/a2810d419c9de0615add1950fb0cbdaecaa54680) was *not* blocked. I believe this is because this secret includes a validity check, and the secret I supplied was not valid.

I then created a Docker Personal Access Token (public read-only key) and committed the value. This push failed as expected/hoped:

```
$ git push
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 8 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (4/4), 424 bytes | 424.00 KiB/s, done.
Total 4 (delta 1), reused 0 (delta 0), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
remote: error: GH013: Repository rule violations found for refs/heads/main.
remote: 
remote: - GITHUB PUSH PROTECTION
remote:   —————————————————————————————————————————
remote:     Resolve the following violations before pushing again
remote: 
remote:     - Push cannot contain secrets
remote: 
remote:     
remote:                 GITHUB PUSH PROTECTION
remote:     ——————————————————————————————————————————————————————
remote:      Resolve the following secrets before pushing again.
remote:     
remote:      (?) Learn how to resolve a blocked push
remote:      https://docs.github.com/code-security/secret-scanning/pushing-a-branch-blocked-by-push-protection
remote:     
remote:     
remote:     —— Docker Personal Access Token ——————————————————————
remote:      locations:
remote:        - commit: eeb7f1c253e8031b39d4de174cbbfc8f2658909b
remote:          path: conf/application.conf:5
remote:     
remote:      (?) To push, remove secret from commit(s) or follow this URL to allow the secret.
remote:      https://github.com/civiform/secret-testing/security/secret-scanning/unblock-secret/2fT177DAWLAnxVuKT3XOITSeL1z
remote:     
remote: 
remote: 
To github.com:civiform/secret-testing.git
 ! [remote rejected] main -> main (push declined due to repository rule violations)
error: failed to push some refs to 'github.com:civiform/secret-testing.git'
```

**Comment by gwendolyngoetz** [comment]
Are you able to force it (in the case of false positives)? If you can what happens if it's in a PR when you merge to another branch?

**Comment by yotommy** [comment]
![image](https://github.com/civiform/civiform/assets/1391583/232ae89e-4600-4dbe-b501-b373b75f1e8f)

**Comment by yotommy** [comment]
Alert: https://github.com/civiform/secret-testing/security/secret-scanning/1

**Comment by yotommy** [comment]
> Are you able to force it (in the case of false positives)?

Yes. The error message includes a URL that can be used to override the protection.

> If you can what happens if it's in a PR when you merge to another branch?

I created a branch and overrode the protection to [push](https://github.com/civiform/secret-testing/commit/0302d5e19e6f6c9db66c3e4a085eedf5b7cf2eda) a new secret to the branch.

I then created a [PR](https://github.com/civiform/secret-testing/pull/1) to merge the change to `main`. It merged without an error.

**Comment by yotommy** [comment]
I was able to push a [commit](https://github.com/civiform/secret-testing/commit/127300e5871166324d96f208d095b40e52eb1141) that contained a throwaway Auth0 client id and secret (they are not on the [list](https://docs.github.com/en/enterprise-cloud@latest/code-security/secret-scanning/secret-scanning-patterns#supported-secrets)). (I rotated the secret, so it is no longer valid.)

**Comment by yotommy** [comment]
GitHub Advanced Security works great *for supported secrets*. Tying the override to GitHub alerts is really nice. It also doesn't require the developers to configure anything. (GitLeaks and Trufflehog require devs to install and configure [`pre-commit`](https://pre-commit.com/). We may be able to enable this automatically in our codespaces, as mentioned [above](https://github.com/civiform/civiform/pull/7176).)

However, GitHub Advanced Security missed an [Auth0 secret](https://github.com/civiform/secret-testing/commit/127300e5871166324d96f208d095b40e52eb1141) that GitLeaks [caught](https://civiform.slack.com/archives/C01QKN2UNMA/p1712610561481859?thread_ts=1712609537.381579&cid=C01QKN2UNMA). (GitLeaks caught it because the name of the variable matched the [`generic-api-key`](https://github.com/gitleaks/gitleaks/blob/57ac4b3dc7f926b4c40882e476a951506675c95a/config/gitleaks.toml#L482-L488) rule.

GitHub Advanced Security supports [custom patterns](https://docs.github.com/en/enterprise-cloud@latest/code-security/secret-scanning/defining-custom-patterns-for-secret-scanning), but apparently not for free, as the setting doesn't appear in the [repo security settings](https://github.com/civiform/secret-testing/settings/security_analysis):

![image](https://github.com/civiform/civiform/assets/1391583/7e895cf9-024a-4566-b5c3-c0a80a0c23bb)

**Comment by gwendolyngoetz** [comment]
Maybe use GHAS and layer on the thing the uses pre-commit so if you have both you're double protected and if you don't installed or forget the pre-commit

**Comment by yotommy** [comment]
1. @nb1701 has enabled GHAS on our repos.
2. I am setting up `gitleaks` as a pre-commit for those who use a codespace. Anyone who creates a new codespace or rebuilds their codespace will pick up the pre-commit hook.
3. I will write documentation for those with existing codespaces (or don't use codespaces at all) to install the pre-commit hook.

**Comment by isabelcshaw** [comment]
@nb1701 are we able to close this?
